### ISSUE TITLE
`out` parameter diagnostics affect uninitialized variables at the _call site_

### ISSUE THREAD

**Issue by MarijnS95** [microsoft/DirectXShaderCompiler#5093]
A recent PR to prevent leaving `out` parameters uninitialized in function calls (https://github.com/microsoft/DirectXShaderCompiler/pull/5047) seems to have escaped its cage and is now also disallowing uninitialized _arguments into_ functions with `out` paramameters. See for example this (IMO totally valid) `Interlocked*` invocation:

```hlsl
        uint original;
        buffer.InterlockedAdd(offset, value, original);
        return original;
```

```
bindless.hlsl:110:46: error: variable 'original' is uninitialized when used here [-Werror,-Wuninitialized]
        buffer.InterlockedAdd(offset, value, original);
                                             ^~~~~~~~
bindless.hlsl:109:22: note: initialize the variable 'original' to silence this warning
        uint original;
                     ^
                      = 0
```

(Compiler from https://github.com/microsoft/DirectXShaderCompiler/commit/667fb773cc2fb40bc31dba6d53e5230a196838ff)

I thought the whole point of `out` parameters _and by extension this diagnostic_ is to know and force that the function will initialize `original` here, and it should thus not need to be pre-initialized by the caller.

_Originally posted by @MarijnS95 in https://github.com/microsoft/DirectXShaderCompiler/issues/5047#issuecomment-1462949361_, cc @llvm-beanz

**Comment by llvm-beanz** [comment]
Just going to drop a full-source example here:

```hlsl
RWByteAddressBuffer buffer;

void interlockWrapper(out uint original) {
  buffer.InterlockedAdd(16, 1, original);
}
```

**Comment by llvm-beanz** [comment]
Something to also be aware of here. `-Wuninitialized` in DXC does cause spurious warnings around struct types too. We also have issue #2494, which is caused by our bad AST hygiene.
